src/org/exist/xmldb/RemoteCollection.java

```diff
@@ -622,36 +622,37 @@ private void uploadAndStore(final Resource res) throws XMLDBException {
                 chunk = new byte[MAX_UPLOAD_CHUNK];
             }
             try {
-                int len;
+
                 String fileName = null;
-                while ((len = is.read(chunk)) > -1) {
-                    final List<Object> params = new ArrayList<>();
-                    if (fileName != null) {
-                        params.add(fileName);
-                    }
+                if (chunk.length > 0) {
+                    int len;
+                    while ((len = is.read(chunk)) > -1) {
+                        final List<Object> params = new ArrayList<>();
+                        if (fileName != null) {
+                            params.add(fileName);
+                        }
 
                     /*
                     Only compress the chunk if it is larger than 256 bytes,
                     otherwise the compression framing overhead results in a larger chunk
                     */
-                    if (len < 256) {
-                        params.add(chunk);
-                        params.add(len);
-                        fileName = (String) xmlRpcClientLease.get().execute("upload", params);
-                    } else {
-                        final byte[] compressed = Compressor.compress(chunk, len);
-                        params.add(compressed);
-                        params.add(len);
-                        fileName = (String) xmlRpcClientLease.get().execute("uploadCompressed", params);
+                        if (len < 256) {
+                            params.add(chunk);
+                            params.add(len);
+                            fileName = (String) xmlRpcClientLease.get().execute("upload", params);
+                        } else {
+                            final byte[] compressed = Compressor.compress(chunk, len);
+                            params.add(compressed);
+                            params.add(len);
+                            fileName = (String) xmlRpcClientLease.get().execute("uploadCompressed", params);
+                        }
                     }
-                }
-                // Zero length stream? Let's get a fileName!
-                if (fileName == null) {
-                    final byte[] compressed = Compressor.compress(new byte[0], 0);
+                } else {
+                    // Zero length stream? Let's get a fileName!
                     final List<Object> params = new ArrayList<>();
-                    params.add(compressed);
+                    params.add(chunk);
                     params.add(0);
-                    fileName = (String) xmlRpcClientLease.get().execute("uploadCompressed", params);
+                    fileName = (String) xmlRpcClientLease.get().execute("upload", params);
```

Comment:
this makes a java client using the xmlrpc interface not compatible with a different version of the database, correct?